src/builder.ts

```diff
@@ -83,10 +83,21 @@ export default class Builder {
 		// Connect the input stream to the rw stream
 		dup.setWritable(inputStream);
 
-		Bluebird.resolve(this.docker.buildImage(inputStream, buildOpts))
-			.then((res: NodeJS.ReadWriteStream) => {
-				let errored = false;
-				const outputStream = res
+		let streamError: Error;
+		const failBuild = _.once((err: Error) => {
+			streamError = err;
+			dup.destroy(err);
+			return this.callHook(hooks, 'buildFailure', handler, err, this.layers);
+		});
```

Comment:
`failBuild` aggregates some common error handling. It's called from 3 locations below. Note the use of the lodash `_.once` to avoid it being called more than once.